net/peer.go

```diff
@@ -873,24 +891,19 @@ func (p *Peer) RemoveP2PCollections(collections []string) error {
 	for _, col := range collections {
 		err = p.server.removePubSubTopic(col)
 		if err != nil {
-			for _, topic := range removedTopics {
-				e := p.server.addPubSubTopic(topic, true)
-				if e != nil {
-					return errors.WithStack(e, errors.NewKV("Cause", err))
-				}
-			}
-			return err
+			return p.rollbackRemovePubSubTopics(removedTopics, err)
 		}
+		removedTopics = append(removedTopics, col)
+	}
+
+	if err = txn.Commit(p.ctx); err != nil {
+		return p.rollbackRemovePubSubTopics(removedTopics, err)
 	}
 
 	// If removing the collection topics succeeds, we add back the collections' documents
 	// to the pubsub topics.
-	for _, col := range collections {
-		c, err := store.GetCollectionBySchemaID(p.ctx, col)
-		if err != nil {
-			return err
-		}
-		keyChan, err := c.GetAllDocKeys(p.ctx)
+	for _, col := range storeCollections {
```

Comment:
suggestion: This block can still fail - suggest adding logic for rollback of the final addPubSubTopic calls, and/or a big comment warning about this, particularly as it may fail after the txn has been committed.

AddP2PCollections has the same issue, but I dont think it would result in any visible end-result data differences (unlike this, where docs will no longer sync).